### ISSUE TITLE
Load: update transfer functions (for hash table structs)

### ISSUE THREAD

**Issue by ellynliu** [uchicago-cs/chiventure#95]
In load.c, write/update functions to take a serialized hash table and convert to game state structs. More elaboration coming

Update: write update functions to take in a serialized hash table and load it into an existing hashtable game structs by adding and removing from hashtable and overlaying existing information.

**Comment by jarvislam1999** [comment]
Hi everyone!

Would anyone on the other loading task @ellynliu @jonahdf like to speak to WDL to see what functions they may have on loading the WDL into a game_t struct for game-state? Also, another question would be the mutable/immutable distinction in their WDL file? 

Any other team member who may want to join this load task with me?

**Comment by jarvislam1999** [comment]
After some mulling through and consideration, we have realized that dealing with hashtable functions and overlaying them onto existing hashtable structs is incredibly difficult and messy for objects like items since they are scattered around different rooms so removing items, adding items into room is incredibly messy and frankly, impossible. One of our only logical approaches is to first allocate a room containing all items so that we can make sure no pointers are lost and we can iterate through them, and that none of our codes accidentally delete any rooms. Or, in our more logical approach, we will create a deep copy of all items so that we can free the old ones freely, which would be the most efficient ways of going about this.

We will try to implement this tonight and let you all know. @mpcsh

**Comment by jarvislam1999** [comment]
Updates: I have written almost everything that is needed for the load function, hash and non-hash alike. We have decided after consideration that we will approach storing all items and all paths in an array before passing them down to smaller functions for processing. While there are certainly workarounds that don't need this method, this form of structure allows for the greatest flexibility in semi-incremental-ness if game state happens to include more immutable features in their structs in the future. 

@jonahdf Would you be able to just add in the path for save_room()? It should be very similar to how I stored the items. @ellynliu Would you be willing to write some commented code on the dest field in save_path? Thanks!!

**Comment by jonahdf** [comment]
I asked general, and it seems as if paths are similar to conditions in that WDL loads them in at the beginning and they are untouched after. Since this is the case, I believe we should remove paths entirely from checkpointing, (proto, save, and load). Does this seem reasonable? @jarvislam1999

**Comment by jonahdf** [comment]
I have sent a pull request to dev containing every load.c function and feature that should be ready for our next merge to master. As mentioned in #90, we have decided to not include conditions/paths in our game save, since these do not change throughout the game. Instead, we will let WDL load them in before calling our function.

Since the last update, we realized our current loading file did not use the load_attribute and load_attribute_value functions, since only load_condition called these. Therefore, we removed these functions and backed them up in checkpointing/dump-it-all (our branch for functions that back up immutable structures, in case we need them later). 

As for the "hash table structs," aka the structs and members that require us to use iterators, we successfully implemented the idea Jarvis had in an earlier comment:
1. Create deep copy array of all rooms/items
2. To load all rooms, iterate through every WDL room and load in serialized room
3. To load all players, iterate through all WDL players, and load in serialized player
4. When loading a room, free all items in the room, and load in the saved items
5. Whenever loading a player, free all items from inventory, and load in the saved items

In the future, we are well equipped to add or remove at will, since we have code that loads pretty much everything specified in game-state.
